daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala

```diff
@@ -136,4 +138,70 @@ private[speedy] object SpeedyTestLib {
       parserParameter: ParserParameters[X]
   ): PureCompiledPackages =
     typeAndCompile(Map(parserParameter.defaultPackageId -> pkg))
+
+  private[lf] object Implicits {
+
+    implicit class AddTestMethodsToMachine(machine: Machine) {
+
+      private[lf] def withWarningLog(warningLog: WarningLog): Machine = {
+        new Machine(
+          sexpr = machine.sexpr,
+          traceLog = machine.traceLog,
+          warningLog = warningLog,
+          loggingContext = machine.loggingContext,
+          compiledPackages = machine.compiledPackages,
+          profile = machine.profile,
+          submissionTime = machine.submissionTime,
+          ledgerMode = machine.ledgerMode,
+        )
+      }
+
+      private[speedy] def withCachedContracts(
+          cachedContracts: (ContractId, CachedContract)*
+      ): Machine = {
+        machine.ledgerMode match {
+          case ledger: OnLedger =>
+            for ((contractId, cachedContract) <- cachedContracts) {
+              ledger.updateCachedContracts(contractId, cachedContract)
+            }
+
+          case OffLedger =>
+        }
+
+        machine
+      }
+
+      private[speedy] def withLocalContractKey(contractId: ContractId, key: GlobalKey): Machine = {
+        machine.ledgerMode match {
+          case ledger: OnLedger =>
+            ledger.ptx = ledger.ptx.copy(
+              contractState = ledger.ptx.contractState.copy(
```

Comment:
Not convinced this is the best way to modify the contract state machine, but it at least allows the scenarios we're interested in testing to be mocked up